### PR TITLE
[RFC] snap-bootstrap: be more robust for recover mode, ignore some errors

### PR DESCRIPTION
**Pull request by anonymouse64**
This is the first take at making recover mode more robust against errors, and
trying to do things anyways even if some things fail on us. This will be
important for robust devices that are still able to get to a recover mode when
things go very very wrong in run mode, and gives snaps that are meant to rescue
the device from recover mode a chance to run and do something about fixing the
device.

This needs tests and obviously needs discussion, but I thought it to be useful to open
before our meeting for folks to take a look at.

### CODE REVIEW

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 1]
```diff
@@ -282,43 +295,96 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
 	}
 
 	// 3. mount ubuntu-data for recovery
+	// start out assuming we have everything and lower expectations as things
+	// fail on us
+	haveHost := true
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:300]: what does host mean here?

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 2]
```diff
@@ -282,43 +295,96 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
...
+	saveDir := boot.InitramfsHostWritableDir
+	if haveHost {
+		// don't do fsck on the data partition, it could be corrupted
+		if err := doSystemdMount(device, boot.InitramfsHostUbuntuDataDir, nil); err != nil {
+			logger.Noticef("failed to mount host ubuntu-data: %v", err)
+			haveHost = false
+		}
+	} else {
+		// we don't have the host data, so we should try to mount ubuntu-save
+		// (if it exists) somewhere else
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:330]: I don't understand this comment,  we mount save to run, no, here?

**Review by pedronis — COMMENTED** [review]
I'm slightly confused by some things here. I'm also worried about how the pile of ifs will make tests painful

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 2 above)
Comment (pedronis) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:330]: mmh, I don't think we need this at all. But we need different logic to open save though

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 3]
```diff
@@ -282,43 +295,96 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
...
+			haveSave = false
+		}
 	}
 
 	// 3.1. mount ubuntu-save (if present)
-	haveSave, err := maybeMountSave(disk, boot.InitramfsHostWritableDir, isDecryptDev, nil)
-	if err != nil {
-		return err
+	if haveSave {
+		haveSave, err = maybeMountSave(disk, saveDir, isDecryptDev, nil)
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:345]: this looks wrong, the 2nd argument to maybeMountSave is the place to find it's key, it's not where we mount save

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 4]
```diff
@@ -282,43 +295,96 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
...
+	// start out assuming we have everything and lower expectations as things
+	// fail on us
+	haveHost := true
+	haveSave := true
 	const lockKeysOnFinish = true
 	device, isDecryptDev, err := secbootUnlockVolumeUsingSealedKeyIfEncrypted(disk, "ubuntu-data", boot.InitramfsEncryptionKeyDir, lockKeysOnFinish)
 	if err != nil {
-		return err
+		haveHost = false
+		if isDecryptDev {
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:306]: Could we use some specific error values instead?

**Review by pedronis — CHANGES_REQUESTED** [review]
some things about save are fairly off here

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 5]
```diff
@@ -282,43 +295,96 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
...
+	if haveHost {
+		// 3.2 verify that the host ubuntu-data comes from where we expect it to
+		// right device
+		matches, err := disk.MountPointIsFromDisk(boot.InitramfsHostUbuntuDataDir, diskOpts)
+		if err != nil {
+			// this is a critical error that should stop the line, because we
+			// are supposed to have the host ubuntu-data, but we are unable to
+			// verify that it is the one we expected, but it supposedly was
+			// already mounted, so it may contain dangerous data we don't want
+			// to be exposed to userspace
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:371]:
```suggestion
			// userspace to be exposed to
```

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 1 above)
Comment (bboozzoo) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:300]: maybe it should be named haveHostData or haveData and then the other `haveHostSave`?

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 6]
```diff
@@ -282,43 +295,96 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
...
+			logger.Noticef("failed to identify ubuntu-data partition to mount: %v", err)
+		}
+		// else no decrypted device and we failed to mount it, so just continue
+		// to recover mode without the host ubuntu-data mount
 	}
 
-	// don't do fsck on the data partition, it could be corrupted
-	if err := doSystemdMount(device, boot.InitramfsHostUbuntuDataDir, nil); err != nil {
-		return err
+	// now mount the host partition (if we have it)
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:320]: .. host ubuntu-data ..

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 7]
```diff
@@ -282,43 +295,96 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
...
+	saveDir := boot.InitramfsHostWritableDir
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:321]: Shouldn't we always try to mount under /run/mnt/ubuntu-save and let the actual snapd to figure out where to mount save?

**Comment by anonymouse64** [comment]
Some of these things are still useful, but in the interest of keeping open PR's low, I will close this and perhaps re-open it later, or otherwise split the useful bits out into different PR's

**Comment by anonymouse64** [comment]
also this largely has been superceded by https://github.com/snapcore/snapd/pull/9592

**Pull request closed without merging**